src/backends/gitlab/API.js

```diff
@@ -0,0 +1,195 @@
+          throw err;
+        }
+      });
+  }
+
+  requestHeaders(headers = {}) {
+    return {
+      ...headers,
+      ...(this.token ? { Authorization: `Bearer ${ this.token }` } : {}),
+    };
+  }
+
+  urlFor(path, options) {
+    const cacheBuster = `ts=${ new Date().getTime() }`;
+    const encodedParams = options.params
+          ? Object.entries(options.params).map(
+            ([key, val]) => `${ key }=${ encodeURIComponent(val) }`)
+          : [];
+    return `${ this.api_root }${ path }?${ [cacheBuster, ...encodedParams].join("&") }`;
+  }
+
+  request(path, options = {}) {
+    const headers = this.requestHeaders(options.headers || {});
+    const url = this.urlFor(path, options);
+    return fetch(url, { ...options, headers })
+    .then((response) => {
+      const contentType = response.headers.get("Content-Type");
+      if (options.method === "HEAD") {
+        return Promise.all([response]);
+      }
+      if (contentType && contentType.match(/json/)) {
+        return Promise.all([response, response.json()]);
+      }
+      return Promise.all([response, response.text()]);
+    })
+    .catch(err => [err, null])
+    .then(([response, value]) => {
+      if (!response.ok) return Promise.reject([value, response]);
+      /* TODO: remove magic. */
+      if (value === undefined) return response;
+      /* OK */
+      return value;
+    })
+    .catch(([errorValue, response]) => {
+      const errorMessageProp = (errorValue && errorValue.message) ? errorValue.message : null;
+      const message = errorMessageProp || (isString(errorValue) ? errorValue : "");
+      throw new APIError(message, response && response.status, 'GitHub', { response, errorValue });
+    });
+  }
+  
+  readFile(path, sha, branch = this.branch) {
+    const cache = sha ? LocalForage.getItem(`gh.${ sha }`) : Promise.resolve(null);
+    return cache.then((cached) => {
+      if (cached) { return cached; }
+      
+      // Files must be downloaded from GitLab as base64 due to this bug:
+      //   https://gitlab.com/gitlab-org/gitlab-ce/issues/31470
+      return this.request(`${ this.repoURL }/repository/files/${ encodeURIComponent(path) }`, {
+        params: { ref: branch },
+        cache: "no-store",
+      }).then(response => this.fromBase64(response.content))
+      .then((result) => {
+        if (sha) {
+          LocalForage.setItem(`gh.${ sha }`, result);
+        }
+        return result;
+      });
+    });
+  }
+  
+  fileExists(path, branch = this.branch) {
+    return this.request(`${ this.repoURL }/repository/files/${ encodeURIComponent(path) }`, {
+      method: "HEAD",
+      params: { ref: branch },
+      cache: "no-store",
+    }).then(() => true).catch((err) => {
+      // TODO: 404 can mean either the file does not exist, or if an API
+      //   endpoint doesn't exist. Is there a better way to check for this?
+      if (err.status === 404) {return false;} else {throw err;}
```

Comment:
You can change this whole block to `.catch(err => (err.status === 404 ? false : Promise.reject(err)))`. I think the 404 check is fine, unless GitLab has something more specific we can check (you can access the whole `response` object as `err.meta`).

Reply:
The problem is that with a HEAD request, we don't actually get the content of the error message back, so we can't tell the difference.